### ISSUE TITLE
[Bug]: race in `hadacore_transform`'s tail warp, last rows % 8 rows read stale shared memory when rows % 8 is 1-4 

### ISSUE THREAD

**Issue by Yunzez** [vllm-project/vllm#57014]
### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>

```text
((vllm-27) ) gpu-a6-7:~$ python -m vllm.collect_env
Collecting environment information...
==============================
        System Info
==============================
OS                           : Red Hat Enterprise Linux release 8.10 (Ootpa) (x86_64)
GCC version                  : (GCC) 8.5.0 20210514 (Red Hat 8.5.0-28)
Clang version                : Could not collect
CMake version                : version 4.3.1
Libc version                 : glibc-2.28

==============================
       PyTorch Info
==============================
PyTorch version              : 2.10.0+cu128
Is debug build               : False
CUDA used to build PyTorch   : 12.8
ROCM used to build PyTorch   : N/A

==============================
      Python Environment
==============================
Python version               : 3.12.12 | packaged by Anaconda, Inc. | (main, Oct 21 2025, 20:16:04) [GCC 11.2.0] (64-bit runtime)
Python platform              : Linux-4.18.0-553.89.1.el8_10.x86_64-x86_64-with-glibc2.28

==============================
       CUDA / GPU Info
==============================
Is CUDA available            : True
CUDA runtime version         : Could not collect
CUDA_MODULE_LOADING set to   :
GPU models and configuration : GPU 0: NVIDIA H100 80GB HBM3
Nvidia driver version        : 590.44.01
cuDNN version                : Could not collect
HIP runtime version          : N/A
MIOpen runtime version       : N/A
Is XNNPACK available         : True

==============================
          CPU Info
==============================
Architecture:        x86_64
CPU op-mode(s):      32-bit, 64-bit
Byte Order:          Little Endian
CPU(s):              96
On-line CPU(s) list: 0-95
Thread(s) per core:  1
Core(s) per socket:  48
Socket(s):           2
NUMA node(s):        2
Vendor ID:           GenuineIntel
CPU family:          6
Model:               143
Model name:          Intel(R) Xeon(R) Platinum 8468
Stepping:            8
CPU MHz:             3800.000
CPU max MHz:         3800.0000
CPU min MHz:         800.0000
BogoMIPS:            4200.00
Virtualization:      VT-x
L1d cache:           48K
L1i cache:           32K
L2 cache:            2048K
L3 cache:            107520K
NUMA node0 CPU(s):   0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42,44,46,48,50,52,54,56,58,60,62,64,66,68,70,72,74,76,78,80,82,84,86,88,90,92,94
NUMA node1 CPU(s):   1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39,41,43,45,47,49,51,53,55,57,59,61,63,65,67,69,71,73,75,77,79,81,83,85,87,89,91,93,95
Flags:               fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities

==============================
Versions of relevant libraries
==============================
[pip3] flashinfer-python==0.5.2
[pip3] numpy==2.2.6
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cudnn-frontend==1.22.1
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-cufile-cu12==1.13.1.3
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-cutlass-dsl==4.4.2
[pip3] nvidia-cutlass-dsl-libs-base==4.4.2
[pip3] nvidia-ml-py==13.595.45
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvshmem-cu12==3.4.5
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] pyzmq==27.1.0
[pip3] torch==2.10.0+cu128
[pip3] torchaudio==2.9.0
[pip3] torchvision==0.25.0+cu128
[pip3] transformers==4.57.6
[pip3] triton==3.6.0
[conda] Could not collect

==============================
         vLLM Info
==============================
ROCM Version                 : Could not collect
vLLM Version                 : 0.27.1
vLLM Build Flags:
  CUDA Archs: Not Set; ROCm: Disabled
GPU Topology:
  	GPU0	NIC0	CPU Affinity	NUMA Affinity	GPU NUMA ID
GPU0	 X 	NODE	0,2,4,6,8,10	0		N/A
NIC0	NODE	 X

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_0

==============================
     Environment Variables
==============================
PYTORCH_NVML_BASED_CUDA_CHECK=1
TORCHINDUCTOR_COMPILE_THREADS=1
TORCHINDUCTOR_CACHE_DIR=/tmp/torchinductor_user

((vllm-27) ) gpu-a6-7:~$
```

</details>

### 🐛 Describe the bug

 **My environment**

- vLLM 0.27.1 (pip wheel), torch 2.13.0+cu130, RTX 4090 (sm_89); also reproduced on H100 (sm_90) with vLLM 0.25.1 and 0.27.1, and on A100-40GB (sm_80) with 0.27.1.
- The kernel source is unchanged on `main` as of 2026-09-14 (`csrc/libtorch_stable/quantization/hadamard/hadacore/hadamard_transform_cuda.cu`, last touched by #42339).

`torch.ops._C.hadacore_transform` returns wrong, run-to-run different values in the last `rows % 8` rows of its input whenever the input has more than 8192 rows and `rows % 8` is 1, 2, 3 or 4. Inputs with at most 8192 rows, or a remainder of 0, 5, 6 or 7, are correct. fp16 and bf16, d = 32 / 64 / 128 alike.

```
rows=  8192 rows%8=0: run-to-run differing rows []              max err vs exact 0.06
rows=  8193 rows%8=1: run-to-run differing rows [8192]          max err 10.67
rows= 20004 rows%8=4: run-to-run differing rows [20000..20003]
rows= 20005 rows%8=5: run-to-run differing rows []
rows= 65540 rows%8=4: run-to-run differing rows [65536..65539]  max err 10.86
```

repro: [repro_hadacore_tail_rows.py](https://gist.github.com/Yunzez/1f812fc3a124d71d744e00e3b153c370#file-repro_hadacore_tail_rows-py)

**Mechanism** (read from the source, row rule and nondeterminism measured). The kernel works in warps of `num_chunks` 256-element chunks (two chunks of 8 rows at d = 64). The copy loop always issues and commits `num_chunks` `cp.async` groups per warp. When the padded chunk count is odd, the last warp owns a single real chunk (`real_num_chunks = 1`, `diff_num_chunks = -1`). In the masked wait loop the `cp.async.wait_group` is issued through `switch (k + diff_num_chunks)`, which for `k = 0` evaluates to `-1`; there is no such case, so no wait is issued and that warp reads its chunk from shared memory before the asynchronous copy has landed. With few blocks the copy usually wins; with a thousand or more blocks it frequently does not. A one-chunk tail is exactly `rows % 8 in 1..4`.

What the warp reads is whatever the previous kernel left in that SM's shared memory. Inverting the corrupted rows through the exact butterfly recovers those bytes: in 117 of 188 corrupted rows they are, whole, the input of the *previous* `hadacore_transform` launch (a constant-row tensor launched just before, at a correct shape); in 19 more they are a mix of that tensor and the row's own data at 16-byte (`cp.async` segment) granularity, i.e. the copy lands while the warp is already reading.

The int4 per-token-head KV mode rotates Q with this kernel on every attention call (rows = tokens × num_query_heads) and K/V on every cache write (rows = tokens × num_kv_heads). With Qwen2.5-0.5B-Instruct (14 query heads, 2 KV heads), single prompt, greedy, `LLM.generate(..., logprobs=1)`, 5 runs each:

```
vllm serve Qwen/Qwen2.5-0.5B-Instruct --kv-cache-dtype int4_per_token_head   # TRITON_ATTN is selected
T=1000 tokens (Q rows 14000, %8=0): first token identical in all runs
T=1002 tokens (Q rows 14028, %8=4): token 0 (logprob -11.9) or token 1023 with logprob NaN; 3 distinct outputs
T=1003 tokens (Q rows 14042, %8=2): same
T=1004 tokens (Q rows 14056, %8=0): identical
```

The K/V write path alone reproduces it and the damage persists: with `max_num_batched_tokens=8448` a single 4097- or 4101-token prompt (K/V rows 2T = 8194 / 8202, remainder 2, while Q rows 14T have remainder 6) returns NaN log-probs at every generated position in 5/5 runs, and 4096 / 4100 tokens are bit-identical across runs. The corrupted K/V sits in the cache and every later step (and any prefix-cache hit) reads it. Through the OpenAI server a request that asks for `logprobs` fails with HTTP 400 `Out of range float values are not JSON compliant: nan`; without `logprobs` it returns wrong tokens silently. We first saw it as sporadic 400s under a concurrent workload (the batch's token count decides the remainder). `int8_per_token_head` and the bf16 cache are unaffected (no Hadamard rotation). The compressed-tensors `transform` module also calls `hadacore_transform` and is exposed at the same row counts.

Exposure follows from the head counts: a model is affected on the Q path when `num_query_heads` is not a multiple of 8 (Qwen2.5 0.5B / 1.5B / 7B: 14 / 12 / 28) and on the K/V path when `num_kv_heads` is not (Qwen2.5 0.5B / 1.5B / 3B: 2; 7B: 4); 32/8-head models never produce a bad remainder.

### Reproducer (kernel level, seconds, needs only vLLM and an sm_80+ GPU)

```python
# Standalone reproducer (needs only vLLM): the Hadacore Hadamard-transform kernel
# (`torch.ops._C.hadacore_transform`, used by --kv-cache-dtype int4_per_token_head for the
# randomized Hadamard transform of Q on every attention call and of K/V on every cache write)
# computes the LAST (rows % 8) rows from undefined data whenever rows > 8192 and rows % 8 != 0.
# Those rows differ run to run and are wrong against the exact transform.  Rows <= 8192, or a
# multiple of 8, are correct.
#
#   python3 repro_hadacore_tail_rows.py          # any CUDA GPU with sm_80+
import torch
from vllm import _custom_ops as ops

def exact(x):                       # unnormalized Walsh-Hadamard butterfly in fp32, then /sqrt(d)
    rows, d = x.shape; y = x.float(); h = 1
    while h < d:
        v = y.view(rows, d // (2 * h), 2, h); y = torch.stack([v[..., 0, :] + v[..., 1, :], v[..., 0, :] - v[..., 1, :]], -2).reshape(rows, d); h <<= 1
    return (y / d ** 0.5).to(x.dtype)

torch.manual_seed(0)
d = 64
for rows in (8192, 8193, 8200, 8201, 16385, 65537, 65540, 70812):
    x = (torch.randn(rows, d, device="cuda") * 3).bfloat16()
    y1 = ops.hadacore_transform(x.clone(), inplace=True)
    y2 = ops.hadacore_transform(x.clone(), inplace=True)
    nondet = torch.nonzero((y1 != y2).any(dim=1)).flatten().tolist()
    err = (y1.float() - exact(x).float()).abs().amax(dim=1)
    wrong = torch.nonzero(err > 1.0).flatten().tolist()
    print(f"rows={rows:6d} rows%8={rows % 8}: run-to-run differing rows {nondet[:4]}{'...' if len(nondet) > 4 else ''}  "
          f"rows with |err|>1 vs exact {wrong[:4]}{'...' if len(wrong) > 4 else ''}  max err {float(err.max()):.2f}")
```

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by he-yufeng** [comment]
Verified the mechanism against current main (`hadamard_transform_cuda.cu`):

- The commit loop (L136-L152) issues `cp.async` + `commit_group` unconditionally for every `k < num_chunks`, so each warp always commits exactly `num_chunks` groups, tail warp included (padding chunks and all).
- The masked wait loop (L340-L375) reads chunk `k` right after the switch. To have group `k` landed by read time, it needs `cp.async.wait_group(num_chunks - k - 1)`, which is exactly what `case k` emits. Indexing the switch with `k + diff_num_chunks` maps the single-real-chunk tail warp (`real_num_chunks = 1`, `diff_num_chunks = -1`) at `k = 0` to `switch(-1)`: no case, no wait, and the read below runs against shared memory the async group may still be filling. That is precisely the `rows % 8 in 1..4` signature and the previous-launch bytes you recovered.
- So the minimal correct fix is to index the switch with `k` (drop the diff term entirely), not `k + diff_num_chunks`; the always-committed `num_chunks` groups are what the wait arithmetic keys off, not `real_num_chunks`.

This one is squarely your discovery end to end, so: are you already writing the fix? If not I am happy to take it, with your repro adapted into a kernel-level test (tail-warp shapes 8193/20004/65540 plus the multiples-of-8 controls) and the one-term switch change; I cannot run GPU here, so I would verify statically and lean on the GPU CI for the runtime proof.

**Comment by Yunzez** [comment]
@he-yufeng Appreciate the timely response! Albeit I found the issue, I think you are defintely more qualified than me to attempt a fix. It's also hard for me to request a GPU (There is a really long queue, always). I would prefer you attempt a fix, but if you are at capacity, I can give it try too 😃

**Comment by he-yufeng** [comment]
No new fix needed from either of us: this exact race is already fixed by the open PR #53167 (`hadacore: issue cp.async.wait_group on the partial-chunk warp`), which has been sitting unreviewed since 2026-08-20. I just verified its mechanism independently: the `llabs(diff_num_chunks)` form issues the right `wait_group` for both the original always-commit-`num_chunks` behavior and the commit-real-only behavior its new prefetch guard introduces, the `default: wait_group 0` makes the negative-index path unreachable, and the same diff also stops the tail warp's dummy `cp.async` groups from reading past the end of the input. Your tail shapes (`rows % 8 in 1..4`, `rows > 8192`) are exactly the partial-chunk-warp case its new parametrized test covers. The useful move now is getting #53167 reviewed; I left an independent verification comment there with the same analysis.
